services/web/server/src/simcore_service_webserver/tag_handlers.py

```diff
@@ -0,0 +1,83 @@
+
+from .db_models import tags
+from .login.decorators import RQT_USERID_KEY, login_required
+
+
+@login_required
+async def list_tags(request: web.Request):
+    uid, engine = request[RQT_USERID_KEY], request.app[APP_DB_ENGINE_KEY]
+    async with engine.acquire() as conn:
+        columns = [col for col in tags.columns if col.key != 'user_id'] # pylint: disable=not-an-iterable
+        query = sa.select(columns).where(tags.c.user_id == uid)
+        result = []
+        async for row_proxy in conn.execute(query):
+            row_dict = dict(row_proxy.items())
+            result.append(row_dict)
+    return result
+
+
+@login_required
+async def update_tag(request: web.Request):
+    uid, engine = request[RQT_USERID_KEY], request.app[APP_DB_ENGINE_KEY]
+    tag_id = request.match_info.get('tag_id')
+    tag_data = await request.json()
+    async with engine.acquire() as conn:
+        # pylint: disable=no-value-for-parameter
+        query = tags.update().values(
+            name=tag_data['name'],
+            description=tag_data['description'],
+            color=tag_data['color']
+        ).where(and_(tags.c.id == tag_id, tags.c.user_id == uid)).returning(
+            tags.c.id,
+            tags.c.name,
+            tags.c.description,
+            tags.c.color
+        )
+        async with conn.execute(query) as result:
+            if result.rowcount == 1:
+                row_proxy = await result.first()
+                return dict(row_proxy.items())
+            raise web.HTTPInternalServerError()
+
+
+@login_required
+async def create_tag(request: web.Request):
+    uid, engine = request[RQT_USERID_KEY], request.app[APP_DB_ENGINE_KEY]
+    tag_data = await request.json()
+    async with engine.acquire() as conn:
+        # pylint: disable=no-value-for-parameter
+        query = tags.insert().values(
+            user_id=uid,
+            name=tag_data['name'],
+            description=tag_data['description'],
+            color=tag_data['color']
+        ).returning(
+            tags.c.id,
+            tags.c.name,
+            tags.c.description,
+            tags.c.color
+        )
+        async with conn.execute(query) as result:
+            if result.rowcount == 1:
+                row_proxy = await result.first()
+                return dict(row_proxy.items())
+            raise web.HTTPInternalServerError()
+
+
+@login_required
+async def delete_tag(request: web.Request):
+    uid, engine = request[RQT_USERID_KEY], request.app[APP_DB_ENGINE_KEY]
+    tag_id = request.match_info.get('tag_id')
+    async with engine.acquire() as conn:
+        # pylint: disable=no-value-for-parameter
+        query = tags.delete().where(
+            and_(tags.c.id == tag_id, tags.c.user_id == uid)
+        )
+        async with conn.execute(query) as result:
+            if result.rowcount == 1:
+                raise web.HTTPNoContent(content_type='application/json')
+            raise web.HTTPInternalServerError()
```

Comment:
If tag you try to delete does not exist, it raises error ... why? The purpose of deleting something is that it stops to exist ... if it was not there in the first time, as a user, it is also deleted.
I would raise error only if the tag remains still remains there (i.e. cannot delete, )

Reply:
we can do it like that... but I think the delete operation failed if it didn't delete anything... and also, it is not good if I am trying to delete something that doesn't exist.
what would you do then? raise a nocontent no matter the result of executing the query? @pcrespov

Reply:
Yes, as I mentioned above, I would give no error If I delete something that does not exists.